### PR TITLE
Timing metrics for frontend

### PR DESCRIPTION
**Pull request by akshay-krishnan**

### CODE REVIEW

File: gtsfm/evaluation/metrics_report.py [hunk 1]
```diff
@@ -372,7 +372,7 @@ def get_html_header() -> str:
 def generate_metrics_report_html(
     metrics_groups: List[GtsfmMetricsGroup],
     html_path: str,
-    other_pipelines_metrics_groups: Dict[str, List[GtsfmMetricsGroup]],
+    other_pipelines_metrics_groups: Optional[Dict[str, List[GtsfmMetricsGroup]]],
```
Comment (ayushbaid) [inline gtsfm/evaluation/metrics_report.py:375]: Can we just pass an empty dict?

File: gtsfm/runner/gtsfm_runner_base.py [hunk 2]
```diff
@@ -283,10 +288,23 @@ def run(self) -> GtsfmData:
                 self.loader.get_gt_cameras(),
                 gt_scene_mesh=self.loader.get_gt_scene_trimesh(),
             )
+            two_view_estimation_duration_sec = time.time() - two_view_estimation_start_time
 
         i2Ri1_dict, i2Ui1_dict, v_corr_idxs_dict, two_view_reports_dict = unzip_two_view_results(two_view_results_dict)
+        two_view_metrics = two_view_estimator.aggregate_frontend_metrics(
```
Comment (ayushbaid) [inline gtsfm/runner/gtsfm_runner_base.py:294]: can you rename it to `two_view_agg_metrics`?

File: gtsfm/scene_optimizer.py [hunk 3]
```diff
@@ -323,7 +313,7 @@ def save_visualizations(
     plot_ba_input_path: Path,
     plot_results_path: Path,
 ) -> List[Delayed]:
-    """Save SfmData before and after bundle adjustment and camera poses for visualization.
+    """Save SfmData before and after bundle adjustment and camera poses for visualization.i
```
Comment (ayushbaid) [inline gtsfm/scene_optimizer.py:316]: please revert the typo

File: gtsfm/two_view_estimator.py [hunk 4]
```diff
@@ -585,3 +586,55 @@ def apply_two_view_estimator(
...
+    Args:
+        two_view_report_dict: Front-end metrics for pairs of images.
+        images: List of all images for this scene, in order of image/frame index.
+
+    Returns:
+        List of dictionaries, where each dictionary contains the metrics for an image pair.
+    """
+
+    def round_fn(x):
+        return round(x, 2) if x else None
```
Comment (ayushbaid) [inline gtsfm/two_view_estimator.py:606]: Can we use a global variable?

**Review by ayushbaid — APPROVED** [review]
LGTM. Left some minot comments

File: gtsfm/evaluation/metrics_report.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/evaluation/metrics_report.py:375]: Using a mutable data structure (empty dict) as an arg leads to some nasty bugs, so i'd avoid that if possible

File: gtsfm/scene_optimizer.py (hunk 3 above)
Comment (johnwlambert) [reply gtsfm/scene_optimizer.py:316]: +1

File: gtsfm/utils/metrics.py [hunk 5]
```diff
@@ -225,7 +225,7 @@ def compute_rotation_angle_metric(wRi_list: List[Optional[Rot3]], gt_wRi_list: L
         A GtsfmMetric for the N rotation angle errors, in degrees.
     """
     errors = []
-    for (wRi, gt_wRi) in zip(wRi_list, gt_wRi_list):
+    for wRi, gt_wRi in zip(wRi_list, gt_wRi_list):
```
Comment (johnwlambert) [inline gtsfm/utils/metrics.py:228]: @akshay-krishnan could you share your python `black` version?

File: gtsfm/two_view_estimator.py [hunk 6]
```diff
@@ -585,3 +586,55 @@ def apply_two_view_estimator(
...
+
+    Args:
+        two_view_report_dict: Front-end metrics for pairs of images.
+        images: List of all images for this scene, in order of image/frame index.
+
+    Returns:
+        List of dictionaries, where each dictionary contains the metrics for an image pair.
+    """
+
+    def round_fn(x):
```
Comment (johnwlambert) [inline gtsfm/two_view_estimator.py:605]: nit: missing return type annotation

File: gtsfm/evaluation/metrics_report.py (hunk 1 above)
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics_report.py:375]: im just fixing the type hint here, we are already passing a `None` somewhere.

File: gtsfm/two_view_estimator.py (hunk 4 above)
Comment (akshay-krishnan) [reply gtsfm/two_view_estimator.py:606]: I think its overkill

File: gtsfm/utils/metrics.py (hunk 5 above)
Comment (akshay-krishnan) [reply gtsfm/utils/metrics.py:228]: Im using the vscode extension: v2023.4.1

File: gtsfm/bundle/bundle_adjustment.py [hunk 7]
```diff
@@ -349,16 +352,6 @@ def run_ba(
             Valid mask as a list of booleans, indicating for each input track whether it was below the re-projection
                 threshold.
         """
-        logger.info(
-            "Input: %d tracks on %d cameras", initial_data.number_tracks(), len(initial_data.get_valid_camera_indices())
-        )
-        if initial_data.number_tracks() == 0 or len(initial_data.get_valid_camera_indices()) == 0:
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:355]: just curious, why remove this?

Comment (johnwlambert) [reply gtsfm/bundle/bundle_adjustment.py:355]: @akshay-krishnan

File: gtsfm/scene_optimizer.py [hunk 8]
```diff
@@ -423,43 +413,13 @@ def save_gtsfm_data(
...
-
-
 def save_full_frontend_metrics(
     two_view_report_dict: Dict[Tuple[int, int], TwoViewEstimationReport],
     images: List[Image],
     filename: str,
-    matching_regime: ImageMatchingRegime,
     metrics_path: Path,
     plot_base_path: Path,
+    save_retrieval_metrics: bool = True,
```
Comment (johnwlambert) [inline gtsfm/scene_optimizer.py:422]: nice, this is cleaner now

File: gtsfm/runner/gtsfm_runner_base.py [hunk 9]
```diff
@@ -340,3 +361,23 @@ def unzip_two_view_results(
...
+
+def save_metrics_reports(metrics_group_list: List[GtsfmMetricsGroup], metrics_path: str) -> None:
+    """Saves metrics to JSON and HTML report.
+
+    Args:
+        metrics_graph: List of GtsfmMetricsGroup from different modules wrapped as Delayed.
+        metrics_path: Path to directory where computed metrics will be saved.
+
+    Returns:
+        List of delayed objects after saving metrics.
```
Comment (johnwlambert) [inline gtsfm/runner/gtsfm_runner_base.py:374]: doesnt return anything right @akshay-krishnan ?

File: gtsfm/bundle/bundle_adjustment.py (hunk 7 above)
Comment (akshay-krishnan) [reply gtsfm/bundle/bundle_adjustment.py:355]: wanted to print it for each stage, forgot to paste above. did it now.

File: gtsfm/bundle/bundle_adjustment.py [hunk 10]
```diff
@@ -295,6 +298,16 @@ def run_ba_stage_with_filtering(
...
         """Runs bundle adjustment and optionally filters the resulting tracks by reprojection error."""
+        logger.info(
+            "Input: %d tracks on %d cameras", initial_data.number_tracks(), len(initial_data.get_valid_camera_indices())
+        )
+        if initial_data.number_tracks() == 0 or len(initial_data.get_valid_camera_indices()) == 0:
+            # No cameras or tracks to optimize, so bundle adjustment is not possible, return invalid result.
+            logger.error(
+                "Bundle adjustment aborting, optimization cannot be performed without any tracks or any cameras."
+            )
+            return initial_data, initial_data, [False] * initial_data.number_tracks(), 0.0
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:309]: could we add a comment what the last return arg represents? L373 says its for `verbose` but kinda weird its a float

**Review by johnwlambert — APPROVED** [review]
Thank you @akshay-krishnan! LGTM

File: gtsfm/bundle/bundle_adjustment.py (hunk 10 above)
Comment (akshay-krishnan) [reply gtsfm/bundle/bundle_adjustment.py:324]: not sure what you mean, the verbose is the input arg, this float is a return value. It represents the final error after optimization, a dummy in this case. Anyway, I updated the docstring.

**Pull request merged**